File: src/main/java/xyz/phanta/cbtweaker/recipe/LazyAccumulatorMap.java
package xyz.phanta.cbtweaker.recipe;

import io.github.phantamanta44.libnine.util.LazyConstant;
import xyz.phanta.cbtweaker.buffer.BufferGroup;
import xyz.phanta.cbtweaker.buffer.BufferType;

import javax.annotation.Nullable;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Supplier;

public interface LazyAccumulatorMap {

    @Nullable
    Group getGroup(String groupId);

    LazyAccumulatorMap copyAccumulators();

    interface Group {

        <A> Supplier<A> getAccumulator(BufferType<?, A, ?, ?> bufType);

    }

    class Impl implements LazyAccumulatorMap {

        private final Map<String, BufferGroup> bufGroups;
        private final Map<String, BaseGroup> groupCache = new HashMap<>();

        public Impl(Map<String, BufferGroup> bufGroups) {
            this.bufGroups = bufGroups;
        }

        @Nullable
        @Override
        public Group getGroup(String groupId) {
            if (!groupCache.containsKey(groupId)) {
                BufferGroup bufGroup = bufGroups.get(groupId);
                if (bufGroup == null) {
                    groupCache.put(groupId, null);
                    return null;
                } else {
                    BaseGroup group = new BaseGroup(bufGroup);
                    groupCache.put(groupId, group);
                    return group;
                }
            }
            return groupCache.get(groupId);
        }

        @Override
        public LazyAccumulatorMap copyAccumulators() {
            return new CopyingMap(this);
        }

        private static class BaseGroup implements Group {

            private final BufferGroup bufGroup;
            private final Map<BufferType<?, ?, ?, ?>, Supplier<?>> accumCache = new HashMap<>();

            public BaseGroup(BufferGroup bufGroup) {
                this.bufGroup = bufGroup;
            }

            @SuppressWarnings("unchecked")
            @Override
            public <A> Supplier<A> getAccumulator(BufferType<?, A, ?, ?> bufType) {
                return (Supplier<A>)accumCache.computeIfAbsent(bufType, k -> new LazyConstant<>(
                        () -> bufGroup.accumulateIngredients(bufType)));
            }

        }

        private static class CopyingMap implements LazyAccumulatorMap {

            private final LazyAccumulatorMap baseMap;
            private final Map<String, CopyingGroup> groupCache = new HashMap<>();

            public CopyingMap(LazyAccumulatorMap baseMap) {
                this.baseMap = baseMap;
            }

            @Nullable
            @Override
            public Group getGroup(String groupId) {
                if (!groupCache.containsKey(groupId)) {
                    Group baseGroup = baseMap.getGroup(groupId);
                    if (baseGroup == null) {
                        groupCache.put(groupId, null);
                        return null;
                    } else {
                        CopyingGroup group = new CopyingGroup(baseGroup);
                        groupCache.put(groupId, group);
                        return group;
                    }
                }
                return groupCache.get(groupId);
            }

            @Override
            public LazyAccumulatorMap copyAccumulators() {
                return new CopyingMap(this);
            }

            private static class CopyingGroup implements Group {

                private final Group baseGroup;
                private final Map<BufferType<?, ?, ?, ?>, Supplier<?>> accumCache = new HashMap<>();

                public CopyingGroup(Group baseGroup) {
                    this.baseGroup = baseGroup;
                }

                @SuppressWarnings("unchecked")
                @Override
                public <A> Supplier<A> getAccumulator(BufferType<?, A, ?, ?> bufType) {
                    return (Supplier<A>)accumCache.computeIfAbsent(bufType, k -> new LazyConstant<>(
                            () -> bufType.copyAccumulator(baseGroup.getAccumulator(bufType).get())));
                }

            }

        }

    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/util/recipe/LazyAccumulatorMap.kt
package st.evening.mc.cbtweaker.util.recipe

import st.evening.mc.cbtweaker.buffer.BufferGroup
import st.evening.mc.cbtweaker.buffer.BufferGroups
import st.evening.mc.cbtweaker.buffer.BufferType
import st.evening.mc.prelude.api.type.Maybe
import st.evening.mc.prelude.api.type.orNull

interface LazyAccumulatorMap {
    operator fun contains(bufGroupId: String): Boolean

    operator fun get(bufGroupId: String): Group?

    fun copyAccumulators(): LazyAccumulatorMap

    interface Group {
        fun <A> getAccumulator(bufType: BufferType<*, A, *, *>): Lazy<A>
    }

    class Impl(private val bufGroups: BufferGroups) : LazyAccumulatorMap {
        private val groupCache: MutableMap<String, Maybe<BaseGroup>> = mutableMapOf()

        override fun contains(bufGroupId: String): Boolean = bufGroupId in bufGroups

        override fun get(bufGroupId: String): Group? {
            groupCache[bufGroupId]?.let { return it.orNull() }
            val bufGroup = bufGroups[bufGroupId]
            if (bufGroup == null) {
                groupCache[bufGroupId] = Maybe.Nothing()
                return null
            } else {
                val group = BaseGroup(bufGroup)
                groupCache[bufGroupId] = Maybe.Just(group)
                return group
            }
        }

        override fun copyAccumulators(): LazyAccumulatorMap = CopyingMap(this)

        private class BaseGroup(private val bufGroup: BufferGroup) : Group {
            private val accumCache: MutableMap<BufferType<*, *, *, *>, Lazy<*>> = mutableMapOf()

            override fun <A> getAccumulator(bufType: BufferType<*, A, *, *>): Lazy<A> {
                accumCache[bufType]?.let {
                    @Suppress("UNCHECKED_CAST")
                    return it as Lazy<A>
                }
                val accumulator = lazy { bufGroup.accumulateIngredients(bufType) }
                accumCache[bufType] = accumulator
                return accumulator
            }
        }

        private class CopyingMap(private val baseMap: LazyAccumulatorMap) : LazyAccumulatorMap {
            private val groupCache: MutableMap<String, Maybe<CopyingGroup>> = mutableMapOf()

            override fun contains(bufGroupId: String): Boolean = bufGroupId in baseMap

            override fun get(bufGroupId: String): Group? {
                groupCache[bufGroupId]?.let { return it.orNull() }
                val baseGroup = baseMap[bufGroupId]
                if (baseGroup == null) {
                    groupCache[bufGroupId] = Maybe.Nothing()
                    return null
                } else {
                    val group = CopyingGroup(baseGroup)
                    groupCache[bufGroupId] = Maybe.Just(group)
                    return group
                }
            }

            override fun copyAccumulators(): LazyAccumulatorMap {
                return CopyingMap(this)
            }

            private class CopyingGroup(private val baseGroup: Group) : Group {
                private val accumCache: MutableMap<BufferType<*, *, *, *>, Lazy<*>> = mutableMapOf()

                override fun <A> getAccumulator(bufType: BufferType<*, A, *, *>): Lazy<A> {
                    accumCache[bufType]?.let {
                        @Suppress("UNCHECKED_CAST")
                        return it as Lazy<A>
                    }
                    // THE COPY IS LAZY! IF THE ORIGINAL MAP IS MODIFIED, THE COPY'S STATE MAY BECOME INCONSISTENT!!
                    val accumulator = lazy { bufType.copyAccumulator(baseGroup.getAccumulator(bufType).value) }
                    accumCache[bufType] = accumulator
                    return accumulator
                }
            }
        }
    }
}
